src/main.rs

```diff
@@ -75,7 +85,18 @@ async fn main() {
             }
         }
         None => {
-            await_ready(authority.clone(), backoff).await;
+            let ready = await_ready(authority.clone(), backoff);
+            if let Some(timeout) = timeout {
+                if !timeout.is_zero() && tokio::time::timeout(timeout, ready).await.is_err() {
+                    eprintln!(
+                        "linkerd-proxy failed to become ready within {:?} timeout",
+                        timeout
+                    );
+                    std::process::exit(EX_UNAVAILABLE)
+                }
+            } else {
+                ready.await;
+            }
```

Comment:
i still think it might be clearer as:

```suggestion
            let await_timeout = async move {
                if let Some(timeout) = timeout {
                    if !timeout.is_zero() {
                        tokio::time::sleep(timeout).await;
                        return timeout;
                    }
                }
                futures::future::pending().await;
            };

            tokio::select! {
                () = await_ready(authority.clone(), backoff) => {}
                t = await_timeout => {
                    eprintln!(
                        "linkerd-proxy failed to become ready within {:?} timeout",
                        t
                    );
                    std::process::exit(EX_UNAVAILABLE)
                }
           }
```

you could even break out an `async fn await_timeout(timeout: Option<Duration>) { ... }` if you don't want the inline async closure.

I like this because we have a single place where we await_ready instead of having conditional/duplicate awaits; so if this needs to get any more conditions we won't need to duplicate logic, etc.

Not a real blocker, just slightly cleaner imo.